control/control.go

```diff
@@ -688,85 +674,80 @@ func (p *pluginControl) verifyPlugin(lp *loadedPlugin) error {
 	newPlugins := []core.SubscribedPlugin{}
 	var serrs []serror.SnapError
 	for _, r := range requested {
-		// get expanded namespaces from requested metrics
-		newNss, err := p.matchQueryToNamespaces(r.Namespace())
+		// get all metric types available in metricCatalog which fulfill the requested namespace and version (if ver <=0 the latest version will be taken)
+		newMetrics, err := p.metricCatalog.GetMetrics(r.Namespace(), r.Version())
 		if err != nil {
 			log.WithFields(log.Fields{
 				"_block": "control",
 				"action": "expanding-requested-metrics",
 				"query":  r.Namespace(),
+				"err":    err,
 			}).Error("error matching requested namespace with metric catalog")
-			serrs = append(serrs, err)
+			serrs = append(serrs, serror.New(err))
 			continue
 		}
 
 		if controlLogger.Level >= log.DebugLevel {
-			for _, n := range newNss {
+			for _, m := range newMetrics {
 				controlLogger.WithFields(log.Fields{
-					"_block": "control",
-					"ns":     n.String(),
-				}).Debug("Expanded namespaces found")
+					"_block":  "control",
+					"ns":      m.Namespace().String(),
+					"version": m.Version(),
+				}).Debug("Expanded namespaces found in metric catalog")
 			}
 		}
 
-		if len(newNss) > 0 {
-			for _, ns := range newNss {
-				// Get metric types from metric catalog
-				m, err := p.metricCatalog.Get(ns, r.Version())
-				if err != nil {
-					log.WithFields(log.Fields{
-						"_block":    "control",
-						"action":    "expanding requested metrics",
-						"namespace": ns.String(),
-						"version":   r.Version(),
-					}).Error("error retreiving metric given a namespace and version.")
-					serrs = append(serrs, serror.New(fmt.Errorf("error retreiving metric %s:%d", ns.String(), r.Version())))
-					continue
-				}
-				// in case config tree doesn't have any configuration for current namespace
-				// it's needed to initialize config, otherwise it will stay nil and panic later on
-				config := configTree.Get(ns.Strings())
-				if config == nil {
-					config = cdata.NewNode()
-				}
-				newMetrics = append(newMetrics, &metric{
-					namespace: m.Namespace(),
-					version:   m.Version(),
-					config:    config,
-				})
-
-				config = configTree.Get([]string{""})
-				if config == nil {
-					config = cdata.NewNode()
-				}
-				plugin := subscribedPlugin{
-					name:     m.Plugin.Name(),
-					typeName: m.Plugin.TypeName(),
-					version:  m.Plugin.Version(),
-					config:   config,
-				}
+		for _, mt := range newMetrics {
+			// in case config tree doesn't have any configuration for current namespace
+			// it's needed to initialize config, otherwise it will stay nil and panic later on
+			cfg := configTree.Get(mt.Namespace().Strings())
+			if cfg == nil {
+				cfg = cdata.NewNode()
+			}
+			// set config to metric
+			mt.config = cfg
+
+			// loaded plugin which exposes the metric
+			lp := mt.Plugin
+			key := lp.Key()
+
+			// groups metricTypes by a plugin.Key()
+			pmt, _ := newMetricsGroupedByPlugin[key]
```

Comment:
My comment was mostly about how weird it looks (which is true whether it was in snap previously or not) but I don't think it has to be addressed as part of this PR.